### PR TITLE
Fog of war

### PR DESCRIPTION
**Pull request by wellme**
Added a fog of war option to the server
The fog of war is a FE6-12-style fog of war in which you can see the terrain but not enemies. That's pretty much it honestly.

In some cases, the player will do some action that can't be executed because of the presence of an enemy in a foggy tile. These are handled as follows:

- Moving in an occupied tile
The unit bumps against the enemy and is forced to wait
- Summoning/dropping a unit on an occupied tile
Instead of summoning/dropping in the intended spot, the game finds another tile in which the action is valid. If it fails, the game is technically stuck in an infinite loop forever, but that shouldn't be possible as long as the vision range of units is sufficiently large
- Shoving/smiting a unit into an occupied tile
Shoving will fail, resulting in the unit waiting instead. If smiting fails, then the unit attempts to shove. If that fails as well, the unit waits.

Note, this is not ready to be merged as the testing I did isn't enough for me to be certain that this is bug-proof.

### CODE REVIEW

**Comment by SapphireTactician** [comment]
Oooh, this is something I was really looking forward to. I'm sorry I've been unavailable for so long. My health isn't exactly great atm. I got a busy week this week, and next week I'll be unavailable, so I'll try my best to wedge some tests in this week.

File: src/net/fe/FEMultiplayer.java [hunk 1]
```diff
@@ -87,7 +88,7 @@ public static void main(String[] args) {
 			game.init(480, 320, "Fire Emblem Multiplayer");
 			/* Testing code */
 //			game.testFightStage();
-//			game.testOverworldStage();
+			game.testOverworldStage();
```
Comment (rayrobdod) [inline src/net/fe/FEMultiplayer.java]: Short term: comment this back out.

Long term, this project probably needs more main methods.

File: src/net/fe/overworldStage/fieldskill/Shove.java [hunk 2]
```diff
@@ -79,6 +94,22 @@ public static boolean canShove(Grid grid, Unit shover, Unit shovee) {
 		);
 	}
 	
+	public static boolean canShoveWithFog(Grid grid, Unit shover, Unit shovee) {
```
Comment (rayrobdod) [inline src/net/fe/overworldStage/fieldskill/Shove.java]: I'm not saying we *should* assume this, but if smite is two spaces, and the smallest sight is three spaces, wont a unit always be able to see the space that they'd be shoved into?

File: src/net/fe/network/command/MoveCommand.java [hunk 3]
```diff
@@ -39,7 +37,8 @@ public Runnable applyClient(ClientOverworldStage stage, Unit unit, ArrayList<Att
 		
 		return new Runnable() {
 			public void run() {
-				Path p = stage.grid.getShortestPath(unit, unit.getXCoord()+moveX, unit.getYCoord()+moveY);
+				Path p = new Path();
+				p.setNodes(path);
 				stage.grid.move(unit, unit.getXCoord()+moveX, unit.getYCoord()+moveY, true);
```
Comment (rayrobdod) [inline src/net/fe/network/command/MoveCommand.java]: `unit.getXCoord()+moveX` should be the same as `path[path.length - 1].x`; similar for the y coordinate. Using that equivalency all four times, it should be possible to prune the `moveX` and `moveY` variables altogether.

File: src/net/fe/overworldStage/fieldskill/Shove.java (hunk 2 above)
Comment (rayrobdod) [reply src/net/fe/overworldStage/fieldskill/Shove.java]: Right, it's possible to shove enemies. Forgot about that. This is necessary.

**Comment by rayrobdod** [comment]
Larger picture, if a multiplayer game wants to hide something from a player, it should hide that information from the player's client as well. And from what I see, this doesn't do that. I'd only have to use a client with the `Unit::isVisible` method implementation replaced with `return true` to see enemy units. And attack them, although at least that should be catchable with server-side validation.

Granted, I'm not sure how one would be able to do the permission-to-undo-a-movement-if-no-action-selected thing if the client has to ask the server between a movement and an action whether the movement intersects any hidden units due to the trying-to-walk-through-someone-ends-your-turn and client-doesn't-know-where-enemy-units-are things.

**Comment by rayrobdod** [comment]
Dead People See Me. Or, fog still parts around where a friendly unit died. In the image, a blue unit died at the space marked by the cursor, and despite no other blue units nearby, the spaces around the cursor are still cleared of fog.

**Comment by wellme** [comment]
Alright, fixed the issues stated by @rayrobdod, thanks friend.

> Larger picture, if a multiplayer game wants to hide something from a player, it should hide that information from the player's client as well. And from what I see, this doesn't do that. I'd only have to use a client with the Unit::isVisible method implementation replaced with return true to see enemy units. And attack them, although at least that should be catchable with server-side validation.

This would be the sane way of implementing this feature. However, I don't think it's actually possible for the server to send different messages to different clients because sending messages rely on . Even the kick messages are sent to everyone and whichever client has the ID of the message leaves the server.

> Short term: comment this back out.
>
>Long term, this project probably needs more main methods.

Regarding test methods being in the main game code, we could just move them over to the test folder. They were used for JUnit tests but they haven't been modified in two years.

**Comment by rayrobdod** [comment]
In b28b490, the change should have been `unit.getXCoord() + moveX` ⇒ `path[path.length - 1].x`, not `unit.getXCoord() + moveX` ⇒ `unit.getXCoord() + path[path.length - 1].x`. `moveX` had relative coordinates; `path[i].x` has absolute coordinates. Sorry if I didn't make that clear. Since this has been merged, I guess I'll submit a separate PR for that.

**Comment by SapphireTactician** [comment]
@wellme I agree that there should be anti-cheat measure implemented with the fog, but likewise, I'm not exactly sure what the best way of implementing that will be. The honor system will have to do for now, but if some kind of server side check can be performed, that'd be a nice start.

There was some stuff I wanted to comment on, but I'll have to file those as separate reports since it's already merged.

**Pull request merged**